### ISSUE TITLE
akka.http.javadsl Http.get overloaded on return type - can't use in Java

### ISSUE THREAD

**Issue by jon-hanson** [akka/akka-http#2211]
The Akka HTTP introduction :

https://doc.akka.io/docs/akka-http/current/introduction.html

has a code sample:

```
public class HttpServerStreamRandomNumbersTest extends AllDirectives {

  public static void main(String[] args) throws Exception {
    // boot up server using the route as defined below
    ActorSystem system = ActorSystem.create("routes");

    final Http http = Http.get(system);
```

which fails to compile on the last line shown above, as the `Http.get` method is overloaded on the return type. The decompiled Java code is as thus:

```
public class Http implements Extension {

    // ...

    public static Http get(final ActorSystem system) {
        return Http$.MODULE$.get(var0);
    }

    public static Extension get(final ActorSystem system) {
        return Http$.MODULE$.get(var0);
    }
```

(The original Scala code is [here](https://github.com/akka/akka-http/blob/v10.1.5/akka-http-core/src/main/scala/akka/http/javadsl/Http.scala))

This means the code sample won't even compile on Java. As a workaround, the call to `Http.get` can be replaced with `Http$.MODULE$.get`.

The issue is present in V10.1.15, and the `Http` source code appears looks the same on the trunk (master).

**Comment by johanandren** [comment]
Thanks for reporting, this was mentioned earlier and boils down to how the Scala compiler generates the static forwarders, but I can't see that we ever opened an issue to track it here.

https://discuss.lightbend.com/t/http-get-is-ambiguous-from-java/571/13

In that forum topic the person who reported the problems said it could be worked around by installing the Scala plugin for IntelliJ and that there actually only was a problem in the IDE, not when compiling against it with `javac`. So that could be worth checking out.

The Scalac issue here: https://github.com/scala/bug/issues/10812 has been merged for Scala 2.13-M4 (so there is not yet a 2.13 release that will help us fixing this)

**Comment by jon-hanson** [comment]
Actually, you're right in that although Intellij complains about the line in question, the code **does** compile with Maven (and javac). The IDE itself is also able to compile the code. Sorry, that is my mistake - I aassumed when I looked at the decompiled code that the IDE was correct. As such I think it's a non-issue.

**Comment by jrudolph** [comment]
But you are not using Java 11, right? I'd vote for keeping it open just for reference until we release a version build with Scala 2.12.7.

**Comment by jon-hanson** [comment]
> But you are not using Java 11

Not yet... still on Java 8.

Happy to leave it open.

**Comment by SethTisue** [comment]
https://github.com/akka/akka/issues/25449 was closed, should this one be closed too?

**Comment by raboof** [comment]
Right, I think we can close this as fixed in ac1a4f3d4773956f6ab0b503112e256b80b67d1b

**Comment by jlprat** [comment]
We might need to put some emphasis on that Java users should use the 2.12 dependency (which we guess is the one they use, but better be clear about it).

**Comment by johanandren** [comment]
We already do this to some extent (on purpose), for example the maven and gradle boxes here: https://akka.io/docs/ and inside of the docs (here for example: https://doc.akka.io/docs/akka-http/current/introduction.html ) but I don't think we have any text saying explicitly "if you use Java, prefer `_2.12`"

**Comment by 2m** [comment]
We have noticed this in Alpakka as well. Alpakka master currently does not compile with JDK 11, because of the same error:

```
[error] /home/martynas/projects/alpakka/doc-examples/src/main/java/csvsamples/FetchHttpEvery30SecondsAndConvertCsvToJsonToKafkaInJava.java:119:1: reference to get is ambiguous
[error]   both method get(akka.actor.ActorSystem) in akka.http.javadsl.Http and method get(akka.actor.ActorSystem) in akka.http.javadsl.Http match
[error]     Http http = Http.get(system);
```

I tried publishing locally Akka Http master (`10.1.5+49-a0a1d667+20181026-1153`) and compiling Alpakka with that, but still got the same error.

**Comment by 2m** [comment]
False alarm. Coursier did not like my locally published version of akka-http. When tried again with regular dependency resolution, it all is fine in Alpakka.

**Comment by laszlovandenhoek** [comment]
FYI, I currently work around this issue using this little Scala object that provides explicit, unambigous bridge methods:
```
package akka.actor

object ExtensionFactory {
  def bridgeGet[E <: akka.actor.Extension](eId: ExtensionId[E]): ActorSystem => E =
    eId.get

  def Tcp(actorSystem: ActorSystem): akka.stream.javadsl.Tcp =
    bridgeGet(akka.stream.javadsl.Tcp)(actorSystem)

  def Http(actorSystem: ActorSystem): akka.http.javadsl.Http =
    bridgeGet(akka.http.javadsl.Http)(actorSystem)

  def Dns(actorSystem: ActorSystem): akka.io.DnsExt =
    bridgeGet(akka.io.Dns)(actorSystem)

  //add other Extensions to taste
}
```

As a bonus, you can now conjure your `Http` object in Java, using Scala syntax :wink: 

```
import static akka.actorExtensionFactory.Http;
//...
final var http = Http(actorSystem);
```

**Comment by jrudolph** [comment]
Now, hopefully really fixed by updating to Scala 2.12.8 in #2305.
